posthog/async_migrations/migrations/0007_persons_and_groups_on_events_backfill.py

```diff
@@ -0,0 +1,475 @@
+2. Create temporary tables with the relevant columns from `person`, `person_distinct_id`, and `groups`
+3. Copy data from the main tables into them
+4. Optimize the temporary tables to remove duplicates and remove deleted data
+5. Create a dictionary to query each temporary table with caching
+6. Run an ALTER TABLE ... UPDATE to backfill all the data using the dictionaries
+
+Constraints
+===========
+
+1. The migration requires a lot of extra space for the new columns. At least 2x disk space is required to avoid issues while migrating.
+2. We use ZSTD(3) compression on the new columns to save on space and speed up large reads.
+3. New columns need to be populated for new rows before running this async migration.
+"""
+
+TEMPORARY_PERSONS_TABLE_NAME = "tmp_person_0006"
+TEMPORARY_PDI2_TABLE_NAME = "tmp_person_distinct_id2_0006"
+TEMPORARY_GROUPS_TABLE_NAME = "tmp_groups_0006"
+
+# :KLUDGE: On cloud, groups and person tables now have storage_policy sometimes attached
+STORAGE_POLICY_SETTING = lambda: ", storage_policy = 'hot_to_cold'" if settings.CLICKHOUSE_ENABLE_STORAGE_POLICY else ""
+
+# we shouldn't set this too low to avoid false positives for data inconsistency given we sample data
+DEFAULT_ACCEPTED_INCONSISTENT_DATA_RATIO = 0.01
+
+
+class Migration(AsyncMigrationDefinition):
+    description = "Backfill persons and groups data on the sharded_events table"
+
+    depends_on = "0006_persons_and_groups_on_events_backfill"
+
+    posthog_min_version = "1.40.0"
+    posthog_max_version = "1.41.99"
+
+    parameters = {
+        "PERSON_DICT_CACHE_SIZE": (5000000, "ClickHouse cache size (in rows) for persons data.", int),
+        "PERSON_DISTINCT_ID_DICT_CACHE_SIZE": (
+            5000000,
+            "ClickHouse cache size (in rows) for person distinct id data.",
+            int,
+        ),
+        "GROUPS_DICT_CACHE_SIZE": (1000000, "ClickHouse cache size (in rows) for groups data.", int),
+        "TIMESTAMP_LOWER_BOUND": ("2020-01-01", "Timestamp lower bound for events to backfill", str),
+        "TIMESTAMP_UPPER_BOUND": ("2024-01-01", "Timestamp upper bound for events to backfill", str),
+        "TEAM_ID": (
+            None,
+            "The team_id of team to run backfill for. If unset the backfill will run for all teams.",
+            int,
+        ),
+    }
+
+    def precheck(self):
+        # Used to guard against self-hosted users running on `latest` while we make tweaks to the migration
+        if not settings.TEST and not get_instance_setting("ALLOW_EXPERIMENTAL_ASYNC_MIGRATIONS"):
+            return (False, "ALLOW_EXPERIMENTAL_ASYNC_MIGRATIONS is set to False")
+        return analyze_enough_disk_space_free_for_table(EVENTS_DATA_TABLE(), required_ratio=2.0)
+
+    def is_required(self) -> bool:
+
+        # we don't check groupX_created_at columns as they are 0 by default
+        rows_to_backfill_check = sync_execute(
+            """
+            SELECT 1
```

Comment:
Nit: I would have done count() - the LIMIT doesn't make this cheaper I think

Reply:
I've tested this - makes it significantly faster

Also intuitively I'd think count scans the whole table while this stops at the first row it finds that satisfies the condition, which is supported by the testing I did

Reply:
Proof from the playground
<img width="753" alt="Screenshot 2022-09-21 at 15 05 40" src="https://user-images.githubusercontent.com/38760734/191578456-aab8d6f5-5940-49af-9e22-2095fe9dd6e2.png">
<img width="756" alt="Screenshot 2022-09-21 at 15 05 49" src="https://user-images.githubusercontent.com/38760734/191578467-1033d9db-0269-4dc8-a9cd-72f1ad19cb94.png">

Reply:
sent Cloud results over Slack if you're curious